[lldb][DynamicLoader] Fix lldb unable to stop at _dl_debug_state if user set it before the process launched.

ZequanWu opened the pull request:
If user sets a breakpoint at `_dl_debug_state` before the process launched, the breakpoint is not resolved yet. When lldb loads dynamic loader module, it's created with `Target::GetOrCreateModule` which notifies any pending breakpoint to resolve. However, the module's sections are not loaded at this time. They are loaded after returned from [Target::GetOrCreateModule](https://github.com/llvm/llvm-project/blob/0287a5cc4e2a5ded1ae2e4079f91052e6a6b8d9b/lldb/source/Plugins/DynamicLoader/POSIX-DYLD/DynamicLoaderPOSIXDYLD.cpp#L574-L577).  This change fixes it by manually resolving breakpoints after creating dynamic loader module.

ZequanWu committed:
[lldb][DynamicLoader] Fix lldb unable to stop at _dl_debug_state if user set it before the process launched.

jimingham commented:
When a shared library gets loaded, its sections have to have been loaded before asking the breakpoint system to try to set new breakpoints.  Presumably this is how it happens for all the other shared library loads, or breakpoints in shared libraries wouldn't work.

I'm missing why the _dl_debug_state breakpoint is special here, such that it requires a force load of the section info?  How does that happen?

jasonmolenda commented:
Skimming DynamicLoaderPOSIXDYLD (as it is written today), it looks like attach/launch call `SetRendezvousBreakpoint()` which (1) loads ld.so into lldb if it isn't already there, and (2) sets the breakpoint to be notified about future binaries loading.  It loads ld.so via `LoadInterpreterModule` which Finds It somehow (I stopped reading), adds it to the Target, and sets the section load addresses, returns the module_sp.  Then it sets breakpoints on a half dozen notification function names in the ld.so Module.  This is all special-cased for ld.so, and we do not evaluate user breakpoints because we didn't call `Target::ModulesDidLoad` for ld.so.

The code path for other binaries goes through `DynamicLoaderPOSIXDYLD::RefreshModules` which does the normal approach of adding binaries to the Target, setting the load addresses, then calling ModulesDidLoad which will evaluate breakpoints that may need to be set in the new binaries.

It seems like a simpler fix would be to have `DynamicLoaderPOSIXDYLD::LoadInterpreterModule` create a ModuleList with the ModuleSP of ld.so that it has just added to Target and set its section load addresses, then call `ModulesDidLoad` on it so that user breakpoints are evaluated and inserted.  The patch as it is written right now is taking one part of ModulesDidLoad and putting it in a separate method that `DynamicLoaderPOSIXDYLD::LoadInterpreterModule` is calling -- why not just call ModulesDidLoad and delegate this (and possibly other binary-just-loaded) behaviors to it?

ZequanWu commented:
> why not just call ModulesDidLoad and delegate this (and possibly other binary-just-loaded) behaviors to it?

That's what I did first, but it breaks the test `TestModuleLoadedNotifys.ModuleLoadedNotifysTestCase.test_launch_notifications` because of extra broadcaster event. So, I moved out the part of the part just updating breakpoints.

ZequanWu commented:
> I'm missing why the _dl_debug_state breakpoint is special here, such that it requires a force load of the section info? How does that happen?

Jason's comment explains it well. It's because ld.so's loading is special here. Normal binaries are loaded via `DynamicLoaderPOSIXDYLD::RefreshModules` which updates loaded section addresses.

labath commented:
> > why not just call ModulesDidLoad and delegate this (and possibly other binary-just-loaded) behaviors to it?
> 
> That's what I did first, but it breaks the test `TestModuleLoadedNotifys.ModuleLoadedNotifysTestCase.test_launch_notifications` because of extra broadcaster event. So, I moved out the part just updating breakpoints.

The point of the test is to ensure you don't get a hundred notifications, one for each module. Having one notification for ld.so, and 99 for the rest of the modules is ok. It should be fine to just update the test to match new reality.

ZequanWu committed:
format

ZequanWu committed:
Update to use `Target::ModulesDidLoad`.

ZequanWu commented:
> > > why not just call ModulesDidLoad and delegate this (and possibly other binary-just-loaded) behaviors to it?
> > 
> > 
> > That's what I did first, but it breaks the test `TestModuleLoadedNotifys.ModuleLoadedNotifysTestCase.test_launch_notifications` because of extra broadcaster event. So, I moved out the part just updating breakpoints.
> 
> The point of the test is to ensure you don't get a hundred notifications, one for each module. Having one notification for ld.so, and 99 for the rest of the modules is ok. It should be fine to just update the test to match new reality.

Updated to use `Target::ModulesDisLoad` and the test.

jasonmolenda reviewed (approved):
This looks good, thanks for the revisions.

labath commented:
> > > > why not just call ModulesDidLoad and delegate this (and possibly other binary-just-loaded) behaviors to it?
> > > 
> > > 
> > > That's what I did first, but it breaks the test `TestModuleLoadedNotifys.ModuleLoadedNotifysTestCase.test_launch_notifications` because of extra broadcaster event. So, I moved out the part just updating breakpoints.
> > 
> > 
> > The point of the test is to ensure you don't get a hundred notifications, one for each module. Having one notification for ld.so, and 99 for the rest of the modules is ok. It should be fine to just update the test to match new reality.
> 
> Updated to use `Target::ModulesDisLoad` and the test.

Oh, so the problem is that the loaded event gets sent twice (and not that it gets sent as a separate event, as I originally thought). While it's not the end of the world (it appears the mac loaded does something similar as well), this does seem like something that we could prevent. Can you check where does the second event get sent from? Is it by any chance when we go through [this](https://github.com/llvm/llvm-project/blob/main/lldb/source/Plugins/DynamicLoader/POSIX-DYLD/DynamicLoaderPOSIXDYLD.cpp#L460) place ? If so, it should be fairly easy to refactor that code to avoid putting the interpreter module into the event list when it is already loaded.

ZequanWu commented:
> Can you check where does the second event get sent from? Is it by any chance when we go through [this](https://github.com/llvm/llvm-project/blob/main/lldb/source/Plugins/DynamicLoader/POSIX-DYLD/DynamicLoaderPOSIXDYLD.cpp#L460) place ?

The first event is sent when creating the module for ld.so: https://github.com/llvm/llvm-project/blob/458328ae23d318a5055d5bac66426b8551bce01f/lldb/source/Plugins/DynamicLoader/POSIX-DYLD/DynamicLoaderPOSIXDYLD.cpp#L574. It gets notified at here: https://github.com/llvm/llvm-project/blob/458328ae23d318a5055d5bac66426b8551bce01f/lldb/source/Target/Target.cpp#L1660. The second event is sent because I explicitly calls `ModulesDidLoad` in this change (the current version) after the creating the ld.so module.

labath commented:
So, could the fix be as simple as passing notify=false in the first call ?

ZequanWu committed:
disable first notification

ZequanWu commented:
> So, could the fix be as simple as passing notify=false in the first call ?

Yeah, absolutely. Updated.

labath commented on lldb/source/Plugins/DynamicLoader/POSIX-DYLD/DynamicLoaderPOSIXDYLD.cpp:
```diff
@@ -572,9 +572,14 @@ ModuleSP DynamicLoaderPOSIXDYLD::LoadInterpreterModule() {
 
   if (ModuleSP module_sp = target.GetOrCreateModule(module_spec,
-                                                    true /* notify */)) {
+                                                    false /* notify */)) {
```
Might as well change to the [official style](https://llvm.org/docs/CodingStandards.html#comment-formatting): `/*notify=*/false`

jimingham replied:
Can you put a comment in here saying why you pass false here.  The next person who comes to read this won't have the context of this discussion and will likely be confused.

ZequanWu replied:
Updated coding style and added a comment here.

ZequanWu committed:
address comments
